csharp/src/Google.Protobuf/Collections/MapField.cs

```diff
@@ -221,17 +227,15 @@ public bool TryGetValue(TKey key, out TValue value)
             }
         }
 
-        // TODO: Make these views?
-
         /// <summary>
         /// Gets a collection containing the keys in the map.
         /// </summary>
-        public ICollection<TKey> Keys { get { return list.Select(t => t.Key).ToList(); } }
+        public ICollection<TKey> Keys { get { return new MapView<TKey>(this, pair => pair.Key, ContainsKey); } }
```

Comment:
The benefit - 
I think always obtaining the same object once you access property like Keys or Values (or property in general) is just what everyone would expect. And breaking natural expectations is almost always a bad idea.  Let's favor API that behaves in the expected way over saving 16 bytes.